### ISSUE TITLE
Inconsistencies in the API responses [Power Ampache 2]

### ISSUE THREAD

**Issue by icefields** [ampache/ampache#3947]
## Description
While fixing a Power Ampache 2 issue related to Nextcloud with @paulijar, some inconsistencies in the API responses have been noticed by Pauli. We both did some tests, and noticed the issue is reproducible on Nextcloud, on my own Ampache server and even on the demo server.
The API (newest is 6.4 at the time of writing) does not return the correct artist for `action=artist_albums`.
### Describe the bug
I will explain the issue with an example.
When I call `action=artists`, looking at the resulting array, among the artists there is one called "3TEETH" with id 219 and as a separate artist I see "3TEETH, Mick Gordon" with id 194. Then if I call `artist_albums` passing 219, I do not get "3TEETH" with id 219 as the artist, but instead I get this:
```
{"total_count": 1,
"album": [
{
  "artist": {
    "id": "194",
    "name": "3TEETH, Mick Gordon",
    ...
},
"artists": [
  {
    "id": "194",
    "name": "3TEETH, Mick Gordon",
    "prefix": null,
    "basename": "3TEETH, Mick Gordon"
  }
],
```

This unexpected inconsistency is messing up my internal app queries queries, especially in offline mode or while fetching saved data. 
As a workaround, if the right number of albums is not retrieved, in Power Ampache 2 I'm falling back to the network response only (basically removing the local db from the equation). 
In Power Ampache's architecture this violates the single source of thruth pattern. 
As a patch/workaround, if I get more albums from the network than the ones from the query of my internal sqlite db, I just return the network array (does not work in offline mode obviously).
It's ugly, and I don't like anti-patterns.

### To reproduce

This issue is more easily reproducible for featured artists in a specific album (the `artists`, plural, attribute), so easier with "various artists" albums.

1. call `action=artists` and find some artist featured in some multi artist album (should be reproducible with the artist "Athuzela Brown" on the demo server)
2. Use that artist's ID to fetch `artist_albums`
3. notice that the artist with the ID used for the call is not present in the artists array

### Expected behavior
If I make an api call `action=artist_albums` with artist id X, the artist with id X should be present in the results.

### Screenshots
See example in _"Describe the bug"_.

## Environment

* Ampache version: most recent (6.4) and below
* Web server + version: Apache
* Server operating system: debian 12
* Client operating system: android, browser, curl

### Client type
* Android Power Ampache 2
* Web
* Ampache
* SubSonic
* probably DAAP
* probably UPnP
* probably WebDAV

**Comment by lachlan-00** [comment]
Short answer is you have to update your file tags with tags that support multi artist.

The artist tag and album artist tag are strings which can't easily be relied upon for identifying groups of artists. (There isn't a real standard you could rely on for how they're recorded)

Artists and mbid fields are arrays which can be split into individuals.

https://github.com/ampache/ampache/wiki/multi-artist

You should definitely not work around and split these guys into hybrid artists as a client.

It's a server/file problem.

If you don't have tags to help the server identify the split artist names then there isn't an easy way to actually identify they are different.

**Comment by lachlan-00** [comment]
here's an example
![image](https://github.com/ampache/ampache/assets/1305249/78f688ba-836c-4d87-b477-22cc0a6ba4b7)

The artist string is useless. if you split on & you'll miss valid bands in other files so you can't do it all the time

if you split on feat. you're left with an invalid band name as it won't match the actual bands

The last issue is that albumartist tags suck as well but it doesn't have a standard array alternative.

So there is some code in the server to try and identify that from the tags as well as using the mbid the release artist id array if it can't be identified from the other tags

![image](https://github.com/ampache/ampache/assets/1305249/d7ab5d2e-d397-4b61-a4b2-5f76874acc81)

When you update file tags the artists will be split and updated into their own artists and the bad artist id will be cleaned up.

One thing i might do is provide support for a nonstandard tag like adding albumartists to allow updates for people who don't use musicbrainz. This isn't a standard tag though so not something we should always rely on.

![image](https://github.com/ampache/ampache/assets/1305249/3fd4b1f6-c1b9-40a4-9d35-2a9f9392c8bf)

**Comment by icefields** [comment]
I understand this part.
I'm not manipulating anything on the client. The bug is in the fact that when I request the albums from an artist, and then I get those albums, the artist in not present at all in that album object.
ie. If I call `action=album_artists&filter=219` I get this. There's no artist with ID 219 here:

```
{"total_count": 1,
"album": [
{
  "artist": {
    "id": "194",
    "name": "3TEETH, Mick Gordon",
    ...
},
"artists": [
  {
    "id": "194",
    "name": "3TEETH, Mick Gordon",
    "prefix": null,
    "basename": "3TEETH, Mick Gordon"
  }
],
```

**Comment by icefields** [comment]
The single source of data for Power Ampache is the internal SQLite database. 
Every time PA2 gets fresh data from the server (every request pretty much), that new data replaces the old one, lazily updating the database while using the app.
The sqlite internal database is used for multiple reasons, beside its main use as the source of data, it's used for caching and for offline-mode. 
When PA2 makes a request, the UI is listening to an observable that, in short:
- emits data from the sqlite database first, to quickly populate the UI without having to wait for a network response. (Data from the database is picked up using a simple query, something like SELECT * FROM artists).
- internally the observable makes the api call and saves the new data in the database (without emitting!).
- Since theoretically there should only be a single source of data, returning data directly from the network request would be an anti-pattern. So instead, I make the same query as before, but on the newly inserted data, and if there's any change, I emit it to refresh the UI accordingly.

Because of this bug/non-bug, some albums are not associated with the artists I'm querying, so some queries return empty arrays.

**Comment by lachlan-00** [comment]
The album_artists query is using the catalog mapping `src/Repository/AlbumRepository.php`

if it's just returning inconsistent data there the mapping table is weird/somethings up.

it's mostly this query with catalog sorting
```
SELECT DISTINCT `album`.`id`, `album`.`release_type`, `album`.`mbid`
FROM `album`
LEFT JOIN `album_map` ON `album_map`.`album_id` = `album`.`id`
WHERE `album_map`.`object_id` = ?
GROUP BY `album`.`id`, `album`.`release_type`, `album`.`mbid`
```

was this artist 219 in the demo?

**Comment by icefields** [comment]
no, artist 219 is from my own server. You can check the artist called "Athuzela Brown" on the demo server. 
I'm not doing any string manipulation, simply if the right number of album is not retrieved from the db I use the network response. Code is pretty simple:

![image](https://github.com/ampache/ampache/assets/149625124/facad113-6bb7-4558-836f-fd9eddc17d89)

**Comment by icefields** [comment]
the screenshot I attached is basically a patch added to step-3, a couple of comments back:
https://github.com/ampache/ampache/issues/3947#issuecomment-2099593433
Basically, try to return db data, fallback to network data in case of inconsistencies in size. This won't work in offline mode obviously.

**Comment by lachlan-00** [comment]
i see what it is now.

The artist in the album object is the album artist in the album page. 

the artist Athuzela Brown is a song artist and not included in that item.

**Comment by lachlan-00** [comment]
the query is returning all albums where the artist is a song artist / an album artist

the output is only including the album artist in the album object

**Comment by icefields** [comment]
when you say "the query", you're referring to the server, right? 
Shouldn't the call for all artists return only artists that have albums associated with them? 
Ie. If I want `artist_albums` from a certain artist, there should be a reference to the artist in the Album.
As a workaround, if this won't be fixed, I will manually make that association in the internal app database in this fashion:
- call `artist_albums` using artist A
- check if A is referenced in the albums in the response
- if not, add it
- save to db
- emit

**Comment by icefields** [comment]
@paulijar anything you want to add?

**Comment by lachlan-00** [comment]
there are a few things we can do. it depends on what you think it should be.

Technically Athuzela Brown is associated with the album through the song. they are not attached to the album directly.

The album is a collection of songs but the album object has a single album artist (Various Artists)

The server does know that they are related through the song so it will return all albums that the artist appears on.

So Athuzela Brown has 1 album but is not an album artist so the data will not include their id. (right now)

there are 2 easy things to do off the top of my head

* Add album_artist and song artist filters to the artist_albums call. This means if you call Athuzela Brown with the album_Artist filter you will get no results

* Add a songartists array to the album data responses. This will allow linking objects by their song artists as well.

**Comment by paulijar** [comment]
> @paulijar anything you want to add?

I initially believed that in case of collection album from "Various Artists" (or similar), the *song artists* should be listed on the `artists` property of the album results. It's clear that the singular `artist` property refers to the *album artist*. However, I haven't implemented that on Nextcloud Music because it's not quite trivial and I have been busy on other things. But then I noticed that the genuine Ampache server doesn't use the `artists` property that way, either.

So the question is now that would it make any sense to change the behavior so that also song artists would be included the `artists` property of the album. If that doesn't make sense and/or it would conflict with some other requirements of Ampache, then I would suggest that @icefields should do some redesigning with his caching system. For example, would it be possible to define another column "query_artist_id" in your SQLite DB which would store the artist ID passed as the argument of `action=artist_albums`, separately from the data retrieved from the server?

**Comment by icefields** [comment]
@lachlan-00 whatever solution is the less invasive I suppose. Changing things on my side also works @paulijar , I have a couple of solutions in mind. 
In a normal situation, the client should just get the data from the server without too much manipulation, but obviously in this particular case the server is not designed around the Android application. Even if this gets fixed I'm going to need a fallback solution for previous versions.
What if we pass a flag to `action=artists` to only get album-artists, without including those song-artists that are causing this issue?

**Comment by paulijar** [comment]
@icefields:
> What if we pass a flag to `action=artists` to only get album-artists, without including those song-artists that are causing this issue?

I guess that it depends on what you want to show your users. Both listing only the album artists and listing all the artists have their uses and it varies from client-to-client, which alternative they are using. And some clients, like Ample, actually provide both alternatives as two separate views:
![image](https://github.com/ampache/ampache/assets/8565946/3eebf5d9-8625-4a10-815a-c9535e3ad71d)

However, remember that you have the "Go to artist" option in the context menu of songs, so just hiding the non-album-artists from the Artists view doesn't mean that the users will never see such artists.

**Comment by lachlan-00** [comment]
artists can show album artists only with album_artist=1, this has been around since at least ampache5 somewhere. (i'd have to check the actual version)

I've added the two options above.

7325818e6 add album_artist=1 to artist_albums

with c3fb802d0 i've added songartists to the album objects

so on an album 'artists' are the direct album artists associated with the album. 'songartists' are the artist of the songs in the album.

Ample splits the albums into a separate category "appears on"
![image](https://github.com/ampache/ampache/assets/1305249/e4485139-baf5-497d-be51-ff0cc42958b9)

I'm happy to add and do more things differently with this. The information is there depending on how you want to display it.

I'm starting work on browse to make the api filterable and sortable as well. that work is something i need to finish before i want to release ampache7 and about half the methods need to be done.

**Comment by lachlan-00** [comment]
the split on query went well (by the looks). This was actually a pretty easy extraction compared to search. 12a45270477c8672397ea4a7263fc2b85af11c1f

After i clean all this up i'll be primarily adding functions to the query classes to support more api actions

**Comment by icefields** [comment]
fantastic, thanks for all the options! I will try the new stuff over the weekend

**Comment by icefields** [comment]
As a fallback for versions before this patch, I do the following:
- make the network request to get artist-albums with artistId=X .
- in the albums in the fetched list, check if `artist.id == X`, check if `artists.map { it.id }` (array of ids) contains the id X .
- if the album doesn't contain the artist with id X, add the artist to the list of artists in `artists`
- save to db
It's a bit cleaner because I'm fixing the missing artist at the source instead of doing checks and fallbacks later.

I think I will use `album_artist=1` in the artists call from now on, to omit song artists. 
@paulijar "Go to Artist" will work thanks to the fix I mentioned previously.
I just tested it with my server and with the demo server, it seems to work fine. Aesthetically, actually, my Artist list looks nicer because many of those song artists don't have an image. 

If you guys can think of any other bad implication in using this flag `album_artist=1`, or if you think it's important to always show song artists, let me know. I really don't know how important can that be for an average user since less than 1% of my collection is compilations. 
In the future, when I start working on filters, I could do something similar to what Ample is doing.

**Comment by lachlan-00** [comment]
I think that's the general expectation from what i've seen with clients.

The majority seem to make the assumption that you're only returning album artists.

Personally, I browse in the web UI by album artist as well but I prefer seeing all the related albums when displaying song artists because the artist is attached to the album.

I've made a new issue regarding browses here https://github.com/ampache/ampache/issues/3949

I will document this as well on the api site in a similar way to the advanced_search pages
